Refactor CollapsibleSection to be more generic

fisher-alice opened the pull request:
<!--
  A summary of the change, including any relevant background, motivation, and context.
  If relevant, include a description, screenshots, and/or video of the existing and new behavior.
-->
Currently `CollapsibleSection` is being used in the levelbuilder edit page for AI customizations. I also need collapsible sections for the presentation view of the model card so this PR refactors the component so that it is more generic.

Updates include:
- option to place an icon to the left of the title (this is required by the mock design of the presentation view of the model card).  I have not yet added the option of adding an icon to the right, but that could be in a follow-up if there is interest in this.
- Title text and content are now clickable so they can be used to expand/collapse content.
- Title text changes to teal color hover.

Note that I set the default values for the title `Typography` and the icons used for expanding/collapsing as the values currently used in `EditAiCustomizations` and `FieldSection` so that changes are not needed other than the location of the import.

There is no change in functionality on the `aichat` levelbuilder edit page other than the updates to the text (teal on hover and clickable to expand/collapse section).

AFTER UPDATE:

Below is a screenshot of collapsible sections with icons from the Presentation view in `aichat` lab.
<img width="574" alt="Screenshot 2024-03-29 at 8 49 59 PM" src="https://github.com/code-dot-org/code-dot-org/assets/107423305/c9c81c15-1118-4ae9-9a32-9f3db74a8da4">

As a comparison, here are collapsible sections in the levelbuilder edit page:
<img width="996" alt="Screenshot 2024-03-29 at 8 53 11 PM" src="https://github.com/code-dot-org/code-dot-org/assets/107423305/f4319532-e827-4daf-991d-b3061003cc0a">


## Links

<!--
  Links to relevant external resources; ie, specification documents, Jira tickets, related PRs, Honeybadger errors, etc.
-->

<!--
- spec: []()
- jira ticket: []()
-->

## Testing story

<!--
  Does your change include appropriate tests?
  If so, please describe how the tests included in this PR are sufficient.
  If not, please explain why this change does not need to be tested.
-->

<!-- Other aspects to consider. Delete any sections that are not relevant to your change. -->

## Deployment strategy

## Follow-up work

<!--
  List (ideally with Jira links) any clean-up or technical debt that will be addressed in future work.
-->

## Privacy

<!--
  1.	Does this change involve the collection, use, or sharing of new Personal Data?
  2.	Does this change involve a new or changed use or sharing of existing Personal Data?
-->

## Security

<!-- Link to Jira task(s) where sensitive security issues are discussed privately. -->

## Caching

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

sanchitmalhotra126 commented on apps/src/templates/CollapsibleSection.tsx:
```diff
@@ -0,0 +1,55 @@
+  expandedIcon?: string;
+}> = ({
+  title,
+  titleSemanticTag = 'p',
+  titleVisualAppearance = 'body-one',
```
What do you think of instead passing in the typography class name (`BodyOneText`) in place of these? Curious what @levadadenys thinks is the best way to pass this information in here

fisher-alice replied:
I'll defer to @levadadenys . I initially thought of this approach, but it seems like maybe I'd need to add a type in `typography/types.ts` for the Typography types defined in `TypographyElements.
Something like:
```
type TypographyElement = Heading1 | Heading2 | ... | BodyOneText | ...;
```
Unsure about this. Thanks!

levadadenys replied:
I think @fisher-alice used the best possible approach here from what I see in the code. If we assume title can be rendered with different sematicTag and VisualAppearance - that's how it's intended to be done.

Passing class name instead won't solve different semantic tags and also will only work if we'll be passing typographyModuleStyles[visualAppearance] which might be harder to read and understand, plus className of typography elements are in moduleStyles, meaning we can't just pass a 'body-one' className string since moduleStyles classNames are parsed into some 'coded' classNames.

Two things we might try here I think:
1) 
```
 import {
   SemanticTag as TypographyElementSematicTag,
   VisualAppearance as TypographyElementVisualAppearance,
 } from '@cdo/apps/componentLibrary/typography/types';
 ```
  for better readability
2) If we're sure that semanticTag will always be 'p' - don't pass it as a prop

fisher-alice replied:
Thanks Denys!
I will update to your first suggestion above.
And `semanticTag` will not always be 'p' so will keep it as a prop. For the presentation view of model card, the `semanticTag` is assigned 'h6'.

sanchitmalhotra126 commented on apps/src/templates/CollapsibleSection.tsx:
```diff
@@ -0,0 +1,55 @@
+        <Typography
+          semanticTag={titleSemanticTag}
+          visualAppearance={titleVisualAppearance}
+          className={moduleStyles.fieldTitle}
```
I think we'll probably want to be able to pass in a custom class name as a prop as well?

sanchitmalhotra126 commented on apps/src/templates/CollapsibleSection.tsx:
```diff
@@ -0,0 +1,55 @@
+  VisualAppearance,
+} from '@cdo/apps/componentLibrary/typography/types';
+
+const CollapsibleSection: React.FunctionComponent<{
```
nit: I know I had put the prop type definition inline (due to hastiness 🙂), but would you mind moving this out into a `CollapsibleSectionProps` definition?

sanchitmalhotra126 commented on apps/src/templates/collapsible-section.module.scss:
```diff
@@ -0,0 +1,17 @@
+@import '@cdo/apps/mixins.scss';
+
+.titleRow {
```
nit: might need to format this file (in case the linter complains)

sanchitmalhotra126 reviewed (commented):
Thanks for doing this! A few minor questions/comments

fisher-alice commented:
> Thanks for doing this! A few minor questions/comments

Hey Sanchit, I addressed your comments but wanted to note that I did make a few updates since you last reviewed since I'm concurrently working on presentation view model card.

I added:
- option to place an icon to the left of the title,
- title text and content are now clickable so they can be used to expand/collapse content,
- title text changes to teal color hover.
I updated the PR description accordingly. Thanks!

sanchitmalhotra126 commented on apps/src/templates/CollapsibleSection.tsx:
```diff
@@ -0,0 +1,84 @@
+  titleStyle?: string;
+  titleIcon?: string;
+  titleIconStyle?: string;
+  children: React.ReactNode;
```
minor readability nit: since there are so many optional props in this, could we move up `children` under `title` so the required props are at the top?

sanchitmalhotra126 commented on apps/src/templates/CollapsibleSection.tsx:
```diff
@@ -0,0 +1,84 @@
+      <div className={moduleStyles.titleRow}>
+        <button
+          type="button"
+          onClick={() => setCollapsed(!collapsed)}
```
nit: pull this out into a `useCallback` and reuse (since it looks like it's used in a few different places)

sanchitmalhotra126 reviewed (approved):
LGTM! Just a few minor comments
